src/entity/Living.php

```diff
@@ -438,6 +444,127 @@ public function setOnFire(int $seconds) : void{
+	 */
+	public function getFreezeProgressTicks() : int{
+		return $this->freezeProgressTicks;
+	}
+
+	/**
+	 * Set the entity's freeze progress in ticks.
+	 *
+	 * @throws \InvalidArgumentException if $freezeProgressTicks is negative
+	 */
+	public function setFreezeProgressTicks(int $freezeProgressTicks) : void{
+		if($freezeProgressTicks < 0){
+			throw new \InvalidArgumentException("Freeze ticks cannot be negative");
+		}
+		$this->freezeProgressTicks = $freezeProgressTicks;
+		$this->networkPropertiesDirty = true;
+	}
+
+	/**
+	 * Returns freeze progress as a normalized value between 0.0 and 1.0.
+	 * This is the value sent to clients to control the freezing visual effect.
+	 */
+	public function getFreezeProgressRatio() : float{
+		return min(1.0, $this->freezeProgressTicks / max(1, $this->getFreezeThresholdTicks()));
+	}
+
+	/**
+	 * Returns the number of ticks required for this entity to be considered fully frozen.
+	 */
+	public function getFreezeThresholdTicks() : int{
+		return 140;
+	}
+
+	/**
+	 * Returns whether the entity is currently flagged as accumulating freeze progress (transient).
+	 */
+	public function isAccumulatingFreeze() : bool{
+		return $this->isAccumulatingFreeze;
+	}
+
+	/**
+	 * Sets the transient freezing flag. This flag indicates that the entity should accumulate freeze progress during
+	 * the next run of `entityBaseTick()`.
+	 *
+	 * @param bool $accumulating Whether to start accumulating freeze progress
+	 */
+	public function setAccumulatingFreeze(bool $accumulating) : void{
+		$this->isAccumulatingFreeze = $accumulating;
+	}
+
+	/**
+	 * Returns whether freeze ticking is currently locked for this entity.
+	 */
+	public function isFreezeTickingLocked() : bool{
+		return $this->freezeLocked;
+	}
+
+	/**
+	 * Locks or unlocks freeze ticking for this entity. When locked, the entity will not accumulate or lose freeze
+	 * progress.
+	 *
+	 * @param bool $locked Whether to lock freeze ticking
+	 */
+	public function lockFreezeTicks(bool $locked = true) : void{
+		$this->freezeLocked = $locked;
+	}
+
+	/**
+	 * Whether this entity can be frozen (i.e. accumulate freeze progress from environments such as powder snow).
+	 */
+	public function isFreezable() : bool{
+		foreach($this->armorInventory->getContents() as $item){
+			if($item instanceof Armor && $item->getMaterial() === VanillaArmorMaterials::LEATHER()){
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Hook called when freeze-based movement modifier should be (re)applied.
+	 */
+	protected function onFreezeAttributeModifierChanged(float $addValue) : void{
+		$base = $this->moveSpeedAttr->getDefaultValue();
+		$oldAdd = $this->freezeMovementAdd;
+		$denom = max(1e-6, $base + $oldAdd);
+		$multiplier = $this->getMovementSpeed() / $denom;
+		$target = ($base + $addValue) * $multiplier;
+
+		$this->setMovementSpeed(max(0.0, $target));
+		$this->freezeMovementAdd = $addValue;
```

Comment:
This function is giving me a headache. What's it trying to do?

Reply:
I'm not convinced that will have the desired effect. You don't know if the current movement speed might've been modified by something else. For example sprinting